### ISSUE TITLE
Synths with multiple organizational entities - how to deal with what?

### ISSUE THREAD

**Issue by markusschloesser** [christofmuc/KnobKraft-orm#58]
my FS1R has 
1. Performances (Presets 384 (128 x 3 banks), Internal 128)
2. Voices (Preset 1408 (128 x 11 banks), Internal 128 / 64 (Internal Fseq))
3. Fseq (Formant Sequence) (Preset 90, Internal 6 (max))

Where each performance consists of 4 Parts (Voices! But they call them parts in this example) and ONE Formant Sequence (and FX).
How would I ideally deal with this on an organizational basis within knobkraft?
It's a bit like the upper/lower concept of the mks80 and others, but more complex.

**Comment by Andy2No** [comment]
I've been wondering the same about my Yamaha TX7, TX802, TX81Z and FB-01 - they all have, at least, banks and performances, which are both a type of preset, but with different formats.

I guess one way would be to regard each one as being at least two different synths, e.g. one synth definition for TX7-patches and one for TX7-performances, each with its own definition of how to store or retrieve a preset.  

The TX802 could also have a third definition for user scales, but I probably wouldn't bother to add that.

Both would have the same response to auto detection, though - because it's the same synth.  I'm not sure if that would cause any problems for Orm.

**Comment by christofmuc** [comment]
Regarding problem 1, multiple file types: 

In C++, this is already possible, it is called the "DataFileLoadCapability" and implements the idea of a Data File type. Currently it is used in the Kawai K3 to distinguish the user wave (additive oscillator) from the patch, and in the Rev2 where you can store Scale files (microtonal scales) independent of the Patches. In the advanced search, there is a "type" dropdown for that purpose.

It is a rather simple exercise to expose this C++ class, now that we have the new dynamic Capability mechanism in the GenericAdaptation - just the Python code to be written will have to be a bit more involved. I can make an example.

I don't think all types need to be supported, e.g. the MKS80 has upper/lower tones, which is useful, but archiving the pesets, the combination of upper/lower, independently of a full patch (preset + upper + lower) is IMHO not really helpful in a librarian. That's something you wish to quickly setup in a good editor. But the adaptation author will have the possibility to implement it, of course.

Regarding problem 2 - nested data types. I don't have a good solution for this yet. Simple example: The Prophet Rev2 patch consists of two layers, which are absolutely identical in structure to earch other (they even both have names, the Rev2 has no Patch or Preset type to bind two layers), think of it as two synth instances in one. 

The straightforward solution: on importing a patch consisting of two layers adding both layers independently as a new data type "layer" into the database. Then you can switch the type filer to layer and see all upper/lower layers, and can click them to send them into a synth layer of your choosing (send to upper/lower). Patches then become just a baked combination of upper/lower.

Why I am reluctant to go with this solution is the potential data duplication - on the other hand, as the database has the concept of identity in that it wants all parameters to be identical to not create a new entry, as soon as you modify either layer you end up with a new layer and a new patch anyway. 

I would rather not go down tracing dependencies between patches, tones, partials etc., that becomes quickly complex when writing into a little SQL database, and doesn't scale well. So I think this needs to be though from the UI experience the user expects, not so much from the data structure.

Regarding problem 3 - the Yamaha synth's format. There is, in C++, a StreamLoadCapability that is meant for the case that one database item is assembled from multiple MIDI messages. The RefaceDX implementation in C++ uses this. The reason I have not exposed it to Python yet is that the abovementioned DataFileLoadCapability (+DataFileSendCapability) is a superset, and I thought it might be confusing if two very similar Capabilities exist (the DataFile just carrying this extra "type" everywhere).

Oh, and @Andy2No  is right that the quick hack solution to the FS1R program is to just create three adaptations, one for each type ;-) At least that could get you going, and the code will be reusable for the new Capability. Autodetection will work as I do it adaptation by adaption - that let's me test my C++ Matrix 1000 code at the same time with the example adaptation for the Matrix 1000.

At some point, the C++ synth implementations should go away, once we can do all we need in Python. They are just much more a hassle to maintain. But we'd need tests for the adaptations, a topic I have been dancing around for awhile now.

**Comment by christofmuc** [comment]
Here is more discussion on the problems of having more than one message per synth, eventually I'll get around to fix it ;-)

https://forums.rolandclan.com/viewtopic.php?f=62&t=59409

**Comment by christofmuc** [comment]
@Andy2No @markusschloesser Coming back to this, now that we have the multiple messages per patch solved, we are still confronted with the multiple types problem. The reason this has not progressed is that I had branches with code in C++, but these quickly got pretty complex and were not good enough to make it into the general release.

I have on the last weekend implemented the transfer for my Behringer RD-9, and that obviously has two different data types, songs and patterns being a drum machine. My solution was just to use the bank and program numbers to make the type known implicitly to the adaptation.

E.g.

RD-9 has 16 songs with 16 patterns per song.

I could just make 16 banks with 17 items each, and know that the first item of each bank will be the song, while the other 16 are patterns.

So:
Program 0 = Song 1
Program 1 = Pattern 1 of Song 1
...
Program 16 = Pattern 16 of Song 1
Program 17 = Song 2
Program 18 = Pattern 1 of Song 2
...
Program 33 = Pattern 16 of Song 2
Program 34 = Song 3
...

This had the advantage to be able to load a full song with all its patterns with a single "bank request" in the Orm.

For other synths, other structures might make more sense, especially when they have a differnt count of each type.

For the adpatation interface extension, we could expand it and say instead of 

def numberOfPatchesPerBank():
   return 16

do rather a 

def getBanks():
   return [ ("Songs", 16), ("Patterns of Song 1", 16,)...)]

which would allow to specify a different number of data entities for each bank.

The rest of the functions could stay the same, we could just add one more like 

def getDatatypes():
   return [ "Song", "Pattern" ]

and add that to the getBanks() function. This is very much like what Sound Diver called BankDrivers.

**Comment by markusschloesser** [comment]
This can be closed as well? With bank descriptors

**Comment by christofmuc** [comment]
Yes, what is missing I think is a proper type display (there is one, but it currently is not Python calculated I think), and then more features could come up, like better support for banks of a certain type and better UI filteres, but we can add these when they become more important.
